cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql

```diff
@@ -0,0 +1,244 @@
+    store.getBlock() = block1 and
+    result.getBlock() = block2 and
+    block1.getASuccessor+() = block2
+  )
+}
+
+/**
+ * Holds if `store` copies the address of `f`'s local variable `var`
+ * into th address `globalAddress`.
+ */
+predicate stackAddressEscapes(
+  StoreInstruction store, StackVariable var, TGlobalAddress globalAddress, Function f
+) {
+  exists(VariableAddressInstruction vai |
+    stackPointerFlowsToUse(store.getSourceValue(), vai) and
+    globalAddress = globalValueNumber(store.getDestinationAddress()) and
+    f = vai.getEnclosingFunction() and
+    var = vai.getASTVariable()
+  ) and
+  // Ensure there's no subsequent store that overrides the global address.
+  not globalAddress = globalValueNumber(getAStoreStrictlyAfter(store).getDestinationAddress())
+}
+
+predicate blockStoresToAddress(
+  IRBlock block, int index, StoreInstruction store, TGlobalAddress globalAddress
+) {
+  block.getInstruction(index) = store and
+  globalAddress = globalValueNumber(store.getDestinationAddress())
+}
+
+predicate blockLoadsFromAddress(
+  IRBlock block, int index, LoadInstruction load, TGlobalAddress globalAddress
+) {
+  block.getInstruction(index) = load and
+  globalAddress = globalValueNumber(load.getSourceAddress())
+}
+
+predicate globalAddressPointsToStack(
+  StoreInstruction store, StackVariable var, CallInstruction call, IRBlock block,
+  TGlobalAddress globalAddress, boolean isCallBlock, boolean isStoreBlock
+) {
+  (
+    if blockStoresToAddress(block, _, _, globalAddress)
+    then isStoreBlock = true
+    else isStoreBlock = false
+  ) and
+  (
+    isCallBlock = true and
+    exists(Function f |
+      stackAddressEscapes(store, var, globalAddress, f) and
+      call.getStaticCallTarget() = f and
+      call.getBlock() = block
+    )
+    or
+    isCallBlock = false and
+    exists(IRBlock mid |
+      mid.immediatelyDominates(block) and
+      // Only recurse if there is no store to `globalAddress` in `mid`.
+      globalAddressPointsToStack(store, var, call, mid, globalAddress, _, false)
+    )
+  )
+}
+
+from
+  StoreInstruction store, StackVariable var, LoadInstruction load, CallInstruction call,
+  IRBlock block, boolean isCallBlock, TGlobalAddress address, boolean isStoreBlock
+where
+  globalAddressPointsToStack(store, var, call, block, address, isCallBlock, isStoreBlock) and
+  block.getAnInstruction() = load and
+  globalValueNumber(load.getSourceAddress()) = address and
+  (
+    // We know that we have a sequence:
+    // (1) store to `address` -> (2) return from `f` -> (3) load from `address`.
+    // But if (2) and (3) happen in the sam block we need to check the
+    // block indices to nsure that (3) happens after (2).
+    if isCallBlock = true
+    then
+      // If so, the load must happen after the call.
+      exists(int callIndex, int loadIndex |
+        blockLoadsFromAddress(_, loadIndex, load, _) and
+        block.getInstruction(callIndex) = call and
+        callIndex < loadIndex
+      )
```

Comment:
If I understand this correctly, this seems rather "coarse" in that it ignores any case in which there is a call after the load, even if the call is unrelated (similarly for the store case below)? I'm not suggestion this should be fixed here and now, if I'm correct 😃

Reply:
That's a good point. I think your conclusion is correct. I have an idea for how to fix this, but I think I'll delay that fix to a later PR.